mnemonics/src/main/java/org/kethereum/bip32/ExtendedKey.kt

```diff
@@ -0,0 +1,257 @@
+            }
+            val mac = Mac.getInstance("HmacSHA512")
+            val key = SecretKeySpec(chainCode, "HmacSHA512")
+            mac.init(key)
+
+            val extended: ByteArray
+            val pub = Keys.getCompressedPublicKey(keyPair)
+            if (isHardened(element)) {
+                val priv = keyPair.privateKey.toBytesPadded(PRIVATE_KEY_SIZE)
+
+                extended = ByteBuffer
+                        .allocate(priv.size + 5)
+                        .order(ByteOrder.BIG_ENDIAN)
+                        .put(0)
+                        .put(priv)
+                        .putInt(element)
+                        .array()
+            } else {
+                //non-hardened
+                extended = ByteBuffer
+                        .allocate(pub.size + 4)
+                        .order(ByteOrder.BIG_ENDIAN)
+                        .put(pub)
+                        .putInt(element)
+                        .array()
+            }
+            val lr = mac.doFinal(extended)
+            val l = Arrays.copyOfRange(lr, 0, PRIVATE_KEY_SIZE)
+            val r = Arrays.copyOfRange(lr, PRIVATE_KEY_SIZE, PRIVATE_KEY_SIZE + CHAINCODE_SIZE)
+
+            val m = BigInteger(1, l)
+            if (m >= CURVE.n) {
+                throw KeyException("Child key derivation resulted in a key with higher modulus. Suggest deriving the next increment.")
+            }
+
+            return if (keyPair.privateKey != BigInteger.ZERO) {
+                val k = m.add(keyPair.privateKey).mod(CURVE.n)
+                if (k == BigInteger.ZERO) {
+                    throw KeyException("Child key derivation resulted in zeros. Suggest deriving the next increment.")
+                }
+                ExtendedKey(ECKeyPair.create(k), r, (depth + 1).toByte(), computeFingerPrint(keyPair), element)
+            } else {
+                val q = CURVE.g.multiply(m).add(CURVE.curve.decodePoint(pub)).normalize()
+                if (q.isInfinity) {
+                    throw KeyException("Child key derivation resulted in zeros. Suggest deriving the next increment.")
+                }
+                val point = CURVE.curve.createPoint(q.xCoord.toBigInteger(), q.yCoord.toBigInteger())
+
+                ExtendedKey(ECKeyPair(BigInteger.ZERO, publicKeyFromPoint(point)), r, (depth + 1).toByte(), computeFingerPrint(keyPair), element)
+            }
+        } catch (e: NoSuchAlgorithmException) {
+            throw KeyException(e)
+        } catch (e: NoSuchProviderException) {
+            throw KeyException(e)
+        } catch (e: InvalidKeyException) {
+            throw KeyException(e)
+        }
+
+    }
+
+    /**
+     * expose keypair
+     */
+    fun getKeyPair(): ECKeyPair {
+        return keyPair
+    }
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (javaClass != other?.javaClass) return false
+
+        other as ExtendedKey
```

Comment:
this looks a bit odd to me

Reply:
This is the auto-generated implementation of equals and hashCode.
It's needed because of the chainCode member that is a ByteArray. It seems the hidden implementation for data classes does not compare ByteArray by contents.

An alternative would be to coerce the chainCode into a BigNumber